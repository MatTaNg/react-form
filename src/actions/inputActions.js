import { userConstants } from '../constants';
import axios from 'axios';

const BASE_URL = 'http://localhost:8080';
export const userActions = {
	login,
	logout,
	register,
	_delete,
	listUsers,
	error,
	resetError
}

function login(userName, password) { 
	let user = { userName: userName, password: password };
	return dispatch => {
		axios.get(`${BASE_URL}/users/${userName}`).then((response) => {
			if(response.data === true) {
				dispatch(resetError());
				dispatch(success( { type: userConstants.LOGIN, user} )); 
			}
			else {
				dispatch(error("Incorrect username or password"))
			}
		}).catch((e) => {
			dispatch(error(e));
		})
	};
	
	function success(user) { return { type: userConstants.LOGIN, payload: user } };
};

function logout() { 
	return dispatch => {
		dispatch(success());
	};
	
	function success() { return { type: userConstants.LOGOUT} };
};

function _delete(userName) {
	return dispatch => {
		axios.delete(`${BASE_URL}/users/${userName}`).then((response) => {
			dispatch(success(response.data));
		}).catch((e) => {
			dispatch(error(e));
		})
	}
	function success(users) { return { type: userConstants.DELETE, users} };
}

function listUsers() {
	return dispatch => {
		axios.get(`${BASE_URL}/users`).then((response) => {
			dispatch(success(response.data));
		}).catch((e) => {
			dispatch(error(e));
		})
	}

	function success(users) { return { type: userConstants.LIST_USERS, users } }
};

function register(user) { 
	return dispatch => {
		return axios.post(`${BASE_URL}/users`, user).then((response) => {
			console.log(response.data)
			if(!response.data) {
				dispatch(error("User already exists"))
			}
			else {
				dispatch(resetError());
				dispatch(success(user));
			}
		}).catch((e) => {
			dispatch(error(e.status));
		})
	};
	function success(user) { return { type: userConstants.REGISTER, payload: user } };
};

function resetError() {
	return { type: userConstants.RESET_ERROR, payload: "" };
};

function error(err) {
	return { type: userConstants.ERROR, payload: err };
};