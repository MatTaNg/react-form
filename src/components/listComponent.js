import React, {PureComponent} from 'react';

export default class ListComponent extends PureComponent {
	render() {
		return (
			<div>
				<ul>
					<li>Test</li>
				</ul>
			</div>
		)
	}
}