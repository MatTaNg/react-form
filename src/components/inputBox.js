import React, { Component } from 'react';
import { addName, editName, deleteName } from './../actions/inputActions';
import { bindActionCreators } from 'redux';
import {connect} from 'react-redux';
import ListComponent from './listComponent';

export class InputBox extends Component {

	constructor(props) {
		super(props);
		this.state = {
			name: "",
			users: []
		};
	}

	componentWillReceiveProps(newProps) {
		this.setState({
			users: newProps.users
		})
	}

	validate = (e) => {
		this.setState({
			name: e.target.value
		});
	};

	add = (e) => {
		this.props.addName(this.state.name);
	}

	edit = (i) => (e) => {
		let newState = [...this.state.users];
		newState[i].edit = true;
		this.setState({
			users: newState
		});
	}

	createUserList = () => {
		let userList = [];
		for(let i = 0; i < this.state.users.length; i++) {
			userList.push(this.userElement(i));
		}
		return userList;
	}

	closeEdit = (i) => () => {
		let newState = [...this.state.users];
		newState[i].edit = false;
		this.setState({
			users: newState
		})
	}

	changeUser = (i) => {
		let newState = [...this.state.users]
		this.props.editName(i, this.state.users[i].name);
		this.closeEdit(i);
	}

	submitChange = (i) => () => {
		let newState = [...this.state.users]
		newState[i].edit = false;
		this.setState({
			users: newState
		})
		this.props.editName(i, this.state.users[i].name);

	}

	delete = (i) => () => {
		let newState = [...this.state.users];
		this.props.deleteName(newState[i].name);
	}

	editUser = (i) => {
		return (
			<div>
				<input type="text" value={this.state.users[i].name}/>
				<input type="submit" onClick={this.submitChange(i)} />
				<span onClick={this.closeEdit(i)}>close</span>
				<span onClick={this.delete(i)}>delete</span>
			</div>
		)
	}

	render() {
		return (
			<div>
	        	<label>Name</label>
      		 	<input type="text" onChange={this.validate} id="name"/>
      		 	<input type="submit" onClick={this.add} value="Add"></input>
      		 	<ul>
      		 		{this.state.users && this.createUserList()}
      		 	</ul>
			</div>
		);
	};

	userElement = (i) => {
		return (
			<div>
				<li key={i}>{this.state.users[i].name}
					<span onClick={this.edit(i)}>Edit</span>
					{ this.state.users[i].edit && this.editUser(i) }
					
				</li>
			</div>
		)
	}
}

const mapStateToProps = (state) => {
	return {
		users: state.users
	}
}

const mapDispatchToProps = (dispatch) => {
	return bindActionCreators({"addName": addName, "deleteName": deleteName, "editName": editName}, dispatch);
}

export default connect(
	mapStateToProps,
	mapDispatchToProps
	)(InputBox)