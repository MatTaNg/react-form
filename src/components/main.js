import React, { Component } from 'react';
import {connect} from 'react-redux';
import { Link, Redirect } from 'react-router-dom'

import { userActions } from '../actions';
import './../styles/main.css';

class Main extends Component {
	constructor(props) {
		super(props);
	}

	componentDidMount () {
		this.props.dispatch(userActions.listUsers())
	}

	handleLogOut = () => {
		this.props.dispatch(userActions.logout());
	};

	deleteUser = id => e => {
		this.props.dispatch(userActions._delete(id));		
	}

	makeUserList = () => {	
		let result = [];
		for(let i = 0; i < this.props.users.length; i++) {
			result.push(
				<li key={i}>{this.props.users[i].userName} 
					<span value={i} className="delete" onClick={this.deleteUser(this.props.users[i].id)}> Delete </span> 
				</li>);
		}
		return this.props.users.length === 0 ? [] : result;
	}

	render() {
		let userList = this.makeUserList();
		if(this.props.user === "") {
			return <Redirect to="/login" />
		}
		return (
			<section className="container">			
				<div className="content content-main">
					<h1 className="heading" >Welcome { this.props.user } ! </h1>
					<ol>
						{ userList }
					</ol>
	  		 		<button className="btn" onClick={this.handleLogOut} >Log Out</button>
  		 		</div>
			</section>
		);
	}
}

function mapStateToProps(state) {
    return {
        user: state.userReducer.loggedInUser,
        users: state.userReducer.users
    };
}

const connectedRegisterPage = connect(mapStateToProps)(Main);
export { connectedRegisterPage as Main };