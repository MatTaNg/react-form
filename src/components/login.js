import React, { PureComponent } from 'react';
import {connect} from 'react-redux';
import { Link, Redirect } from 'react-router-dom'

import { userActions } from './../actions';
import './../styles/main.css';

export class Login extends PureComponent {

	constructor(props) {
		super(props);
		this.props.dispatch(userActions.resetError());

		this.state = {
			username: '',
			password: ''
		};
	}

	login = (e) => {
		e.preventDefault();
		this.props.dispatch(userActions.login(this.state.username, this.state.password));
	}


	handleChange = (e) => {
		const {name, value } = e.target;
		this.setState({ [name]: value });
	}

	render() {
		if(this.props.loggedInUser !== "") {
			return <Redirect to="/main" />
		}
		return (
			<section className="container">

      
				<div className="content">
				
					<h1 className="heading">Login</h1>

					<form className="form">
			        	<label className="label" htmlFor="username">User Name:</label>
		      		 	<input className="input" name="username" type="text" onChange={this.handleChange} id="username"/>

		      		 	<label className="label" htmlFor="password">Password:</label>
		      		 	<input className="input" name="password" type="password" onChange={this.handleChange} id="password"/>

		      		 	<input className="btn float-element" id="login" type="submit" onClick={this.login} value="Login"></input>
	      		 	</form>

	      		 	<Link className="btn" to="/register">Register</Link>

	      		 	<br />
	      		 	<span className="error">{ this.props.error }</span>
      		 	</div>
			</section>
		);
	};

	userElement = (i) => {
		return (
			<div>
				<li key={i}>{this.state.users[i].name}
					<span onClick={this.edit(i)}>Edit</span>
					{ this.state.users[i].edit && this.editUser(i) }
					
				</li>
			</div>
		)
	}
}

const mapStateToProps = (state) => {
	return {
		error: state.userReducer.error,
		loggedInUser: state.userReducer.loggedInUser
	}
}

const connectedLoginPage = connect(
	mapStateToProps
	)(Login);

export default connectedLoginPage;