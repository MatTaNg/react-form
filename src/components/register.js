import React, { PureComponent } from 'react';
import {connect} from 'react-redux';
import { Link } from 'react-router-dom'

import { userActions } from './../actions';
import './../styles/main.css';

 export class Register extends PureComponent {

	constructor(props) {
		super(props);
		this.props.dispatch(userActions.resetError());
		this.state = {
			user: {
				firstName: '',
				lastName: '',
				userName: '',
				password: ''
			}
		}
	}

	handleFormChange = (e) => {
        const { name, value } = e.target;
        const { user } = this.state;
        this.setState({
            user: {
                ...user,
                [name]: value
            }
        });
	}

	submitForm = (e) => {
        e.preventDefault();

        const { user } = this.state;
        const { dispatch } = this.props;
        console.log("USER", this.state);
        if (user.firstName && user.lastName && user.userName && user.password) {
            this.props.dispatch(userActions.register(user));
            if(this.props.error === "") {
	    		this.props.history.push("/login");	
			}
        } else {
        	this.props.dispatch(userActions.error("Form Incomplete"));
        }

	}

	render() {
        const { user } = this.state;
		return (
			<section className="container">


				<div className="content">
					<h1 className="heading">Register</h1>
					<form className="form" name="registrationForm" onSubmit={this.submitForm}>
						<label className="label" htmlFor="firstName">First Name</label>
						<input className="input" type="text" name="firstName" value={user.firstName} onChange={this.handleFormChange} />

						<label className="label" htmlFor="lastName">Last Name</label>
						<input className="input" type="text" name="lastName" value={user.lastName} onChange={this.handleFormChange} />

						<label className="label" htmlFor="userName">User Name</label>
						<input className="input" type="text" name="userName" value={user.userName} onChange={this.handleFormChange} />

						<label className="label" htmlFor="password">Password</label>
						<input className="input" type="password" name="password" value={user.password} onChange={this.handleFormChange} />

						<input className="btn float-element" type="submit" id="register" name="submit" value="Register" />
					</form>
				
	      		 	<Link className="btn" to="/login">Cancel</Link>

	      		 	<br />
	      		 	<span className="error">{ this.props.error }</span>
      		 	</div>
			</section>
		);
	}
}

function mapStateToProps(state) {
    return {
		error: state.userReducer.error
    };
}

const connectedRegisterPage = connect(mapStateToProps)(Register);
export default connectedRegisterPage;