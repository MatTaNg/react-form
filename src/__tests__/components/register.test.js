import React from "react";
import { shallow, mount, render } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";
import { Register } from "./../../components/register";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { userActions } from './../../actions';

// jest.mock('./__mocks__/api');
let wrapper;
let dispatch;
let props;
let userActions_spy; 

describe("Register", () => {

	beforeEach(() => {
		dispatch = jest.fn();
		props = {
			dispatch,
			error: ''
		};
		wrapper = mount(<Router><Register {...props} /></Router>);
	});
	it("renders without crashing", () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});
	it("should dispatch resetError action", () => {
		userActions_spy = jest.spyOn(userActions, 'resetError');
		wrapper = mount(<Router><Register {...props} /></Router>);
		expect(userActions_spy).toHaveBeenCalled();
	});
	// it("should dispatch a register action on form submit", () => {
	// 	let mock_user = { user: {
	// 		firstName: 'mockFN',
	// 		lastName: 'mockLN',
	// 		userName: 'mockUN',
	// 		password: 'mockPW'
	// 	} };
	// 	userActions_spy = jest.spyOn(userActions, 'register');
	// 	wrapper = mount(<Router><Register {...props} /></Router>);		
	// 	wrapper.setState(mock_user, () => {
	// 		console.log("Wrapper", wrapper.instance());
	// 		wrapper.find('#register').simulate('submit'); 
	// 		expect(userActions_spy).toHaveBeenCalled();
	// 	});
	// });
	it("should dispatch an error action if the fields are empty on form submit", () => {
		userActions_spy = jest.spyOn(userActions, 'error');
		wrapper = mount(<Router><Register {...props} /></Router>);		
		wrapper.find('#register').simulate('submit'); 
		expect(userActions_spy).toHaveBeenCalled();
	});
});