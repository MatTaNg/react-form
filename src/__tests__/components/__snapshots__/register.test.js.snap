// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders without crashing 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Register
      dispatch={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "payload": "",
                "type": "RESET_ERROR",
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      error=""
    >
      <section
        className="container"
      >
        <div
          className="content"
        >
          <h1
            className="heading"
          >
            Register
          </h1>
          <form
            className="form"
            name="registrationForm"
            onSubmit={[Function]}
          >
            <label
              className="label"
              htmlFor="firstName"
            >
              First Name
            </label>
            <input
              className="input"
              name="firstName"
              onChange={[Function]}
              type="text"
              value=""
            />
            <label
              className="label"
              htmlFor="lastName"
            >
              Last Name
            </label>
            <input
              className="input"
              name="lastName"
              onChange={[Function]}
              type="text"
              value=""
            />
            <label
              className="label"
              htmlFor="userName"
            >
              User Name
            </label>
            <input
              className="input"
              name="userName"
              onChange={[Function]}
              type="text"
              value=""
            />
            <label
              className="label"
              htmlFor="password"
            >
              Password
            </label>
            <input
              className="input"
              name="password"
              onChange={[Function]}
              type="password"
              value=""
            />
            <input
              className="btn float-element"
              id="register"
              name="submit"
              type="submit"
              value="Register"
            />
          </form>
          <Link
            className="btn"
            to="/login"
          >
            <a
              className="btn"
              href="/login"
              onClick={[Function]}
            >
              Cancel
            </a>
          </Link>
          <br />
          <span
            className="error"
          />
        </div>
      </section>
    </Register>
  </Router>
</BrowserRouter>
`;
